### PR TITLE
Ability caster begin step logic

### PR DESCRIPTION
**Pull request by nick-wilson95**

### CODE REVIEW

**Comment by nick-wilson95** [comment]
[Ability caster begin step logic](https://trello.com/c/F1uZCFpH/10-ability-caster-begin-step-logic)

File: scripts/AbilityCaster_TICK_EFFECTS/AbilityCaster_TICK_EFFECTS.gml [hunk 1]
```diff
@@ -0,0 +1,15 @@
+/// @function AbilityCaster_TICK_EFFECTS() Ticks down active effects, then removes any expired effects.
+
+// Loop through backwards so we check each element exactly once (list length may change throughout this loop).
+for(i=ds_list_size(list_activeEffects)-1; i>=0; i--){
+	// Each element in list_activeEffects has form [effect, arguments, steps_remaining];
+	var _arr_effect = list_activeEffects[| i];
+	// Note, _arr_effect[2] is the number of remaining steps of the effect
+	_arr_effect[2] -= 1;
+	list_activeEffects[| i] = _arr_effect;
```
Comment (Jibzi) [inline scripts/AbilityCaster_TICK_EFFECTS/AbilityCaster_TICK_EFFECTS.gml]: May want to do 0 checking before this line?

File: scripts/AbilityCaster_TICK_EFFECTS/AbilityCaster_TICK_EFFECTS.gml [hunk 2]
```diff
@@ -0,0 +1,15 @@
+/// @function AbilityCaster_TICK_EFFECTS() Ticks down active effects, then removes any expired effects.
+
+// Loop through backwards so we check each element exactly once (list length may change throughout this loop).
+for(i=ds_list_size(list_activeEffects)-1; i>=0; i--){
+	// Each element in list_activeEffects has form [effect, arguments, steps_remaining];
+	var _arr_effect = list_activeEffects[| i];
```
Comment (Jibzi) [inline scripts/AbilityCaster_TICK_EFFECTS/AbilityCaster_TICK_EFFECTS.gml]: Seems a bit wasteful to allocate a new var here. Could this be reworked to directly work off of the list? If not, move this var outside of the for loop, and then set it within.

File: scripts/AbilityCaster_APPLY_EFFECTS_TO_STATS/AbilityCaster_APPLY_EFFECTS_TO_STATS.gml [hunk 3]
```diff
@@ -0,0 +1,3 @@
+/// @function AbilityCaster_APPLY_EFFECTS_TO_STATS() Applies active effects to player stats.
+
+// todo: Cycle through all possible effects. For each, call a script that aggregates all correponding active effects and applies a change to stats.
```
Comment (Jibzi) [inline scripts/AbilityCaster_APPLY_EFFECTS_TO_STATS/AbilityCaster_APPLY_EFFECTS_TO_STATS.gml]: Make sure to make an additional note of this, maybe Trello backlog.

File: objects/oPlayer/Create_0.gml [hunk 4]
```diff
@@ -10,8 +10,8 @@ enum Enum_PlayerMoveModes {
 
 enum_currentMoveMode = Enum_PlayerMoveModes.FREEMOVE;	// Sets the default movement mode to FREEMOVE
 
-// Movement
-num_baseMovementSpeed = 3;
+// Movement todo: make a script for this
```
Comment (Jibzi) [inline objects/oPlayer/Create_0.gml]: Make sure to make an additional note of this, maybe Trello backlog.

File: scripts/AbilityCaster_APPLY_EFFECTS_TO_STATS/AbilityCaster_APPLY_EFFECTS_TO_STATS.gml (hunk 3 above)
Comment (nick-wilson95) [reply scripts/AbilityCaster_APPLY_EFFECTS_TO_STATS/AbilityCaster_APPLY_EFFECTS_TO_STATS.gml]: This is in the sprint

File: objects/oPlayer/Create_0.gml (hunk 4 above)
Comment (nick-wilson95) [reply objects/oPlayer/Create_0.gml]: This comment is void this sprint anyway

File: objects/oAbilityCaster/Step_1.gml [hunk 5]
```diff
@@ -0,0 +1,11 @@
+// Tick down effects, remove if expired
+AbilityCaster_TICK_EFFECTS();
+
+// Add effects from buffer
+AbilityCaster_ADD_BUFFER_EFFECTS();
+
+// Reset current stats to base stats
+map_currentStats = map_baseStats;
```
Comment (Drummerboy444) [inline objects/oAbilityCaster/Step_1.gml]: This isn't copying over properly

File: scripts/AbilityCaster_TICK_EFFECTS/AbilityCaster_TICK_EFFECTS.gml (hunk 1 above)
Comment (nick-wilson95) [reply scripts/AbilityCaster_TICK_EFFECTS/AbilityCaster_TICK_EFFECTS.gml]: Will change

File: scripts/AbilityCaster_TICK_EFFECTS/AbilityCaster_TICK_EFFECTS.gml (hunk 2 above)
Comment (nick-wilson95) [reply scripts/AbilityCaster_TICK_EFFECTS/AbilityCaster_TICK_EFFECTS.gml]: That would be more efficient but I think it's clearer doing it like this?

**Pull request merged**